service/frontend/service.go

```diff
@@ -124,153 +123,115 @@ func NewConfig(dc *dynamicconfig.Collection, numHistoryShards int, enableReadFro
-	return &Service{
-		params: params,
-		config: config,
-		stopC:  make(chan struct{}),
-	}
-}
-
-// Start starts the service
-func (s *Service) Start() {
-
-	var params = s.params
-	var log = params.Logger
-
-	log.Info("starting", tag.Service(common.FrontendServiceName))
+func NewService(
+	params *service.BootstrapParams,
+) (resource.Resource, error) {
 
-	base := service.New(params)
-
-	pConfig := params.PersistenceConfig
-	pConfig.HistoryMaxConns = s.config.HistoryMgrNumConns()
-	pConfig.SetMaxQPS(pConfig.DefaultStore, s.config.PersistenceMaxQPS())
-	pConfig.VisibilityConfig = &config.VisibilityConfig{
-		VisibilityListMaxQPS:            s.config.VisibilityListMaxQPS,
-		EnableSampling:                  s.config.EnableVisibilitySampling,
-		EnableReadFromClosedExecutionV2: s.config.EnableReadFromClosedExecutionV2,
-	}
-	pFactory := client.NewFactory(&pConfig, params.ClusterMetadata.GetCurrentClusterName(), base.GetMetricsClient(), log)
-
-	metadata, err := pFactory.NewMetadataManager()
-	if err != nil {
-		log.Fatal("failed to create metadata manager", tag.Error(err))
-	}
-
-	visibilityFromDB, err := pFactory.NewVisibilityManager()
-	if err != nil {
-		log.Fatal("failed to create visibility manager", tag.Error(err))
+	isAdvancedVisExistInConfig := len(params.PersistenceConfig.AdvancedVisibilityStore) != 0
+	serviceConfig := NewConfig(dynamicconfig.NewCollection(params.DynamicConfig, params.Logger), params.PersistenceConfig.NumHistoryShards, isAdvancedVisExistInConfig)
+
+	params.PersistenceConfig.HistoryMaxConns = serviceConfig.HistoryMgrNumConns()
+	params.PersistenceConfig.SetMaxQPS(params.PersistenceConfig.DefaultStore, serviceConfig.PersistenceMaxQPS())
+	params.PersistenceConfig.VisibilityConfig = &config.VisibilityConfig{
+		VisibilityListMaxQPS:            serviceConfig.VisibilityListMaxQPS,
+		EnableSampling:                  serviceConfig.EnableVisibilitySampling,
+		EnableReadFromClosedExecutionV2: serviceConfig.EnableReadFromClosedExecutionV2,
 	}
 
-	var visibilityFromES persistence.VisibilityManager
-	if params.ESConfig != nil {
-		visibilityIndexName := params.ESConfig.Indices[common.VisibilityAppName]
-		visibilityConfigForES := &config.VisibilityConfig{
-			MaxQPS:                 s.config.PersistenceMaxQPS,
-			VisibilityListMaxQPS:   s.config.ESVisibilityListMaxQPS,
-			ESIndexMaxResultWindow: s.config.ESIndexMaxResultWindow,
-			ValidSearchAttributes:  s.config.ValidSearchAttributes,
-		}
-		visibilityFromES = espersistence.NewESVisibilityManager(visibilityIndexName, params.ESClient, visibilityConfigForES,
-			nil, base.GetMetricsClient(), log)
-	}
-	visibility := persistence.NewVisibilityManagerWrapper(
-		visibilityFromDB,
-		visibilityFromES,
-		s.config.EnableReadVisibilityFromES,
-		dynamicconfig.GetStringPropertyFn(common.AdvancedVisibilityWritingModeOff), // frontend visibility never write
+	serviceResource, err := resource.New(
+		params,
+		common.FrontendServiceName,
+		serviceConfig.ThrottledLogRPS,
+		func(
```

Comment:
can you move this into a separate func or at the very least, a func variable ? This is too much code for an anonymous inline func